Java 21 Gradle DevContainer / Directory

holtkampjs opened the pull request:
# Summary

I wanted to add more targeted DevContainers for interviews. 

## Details

- I switched to `mcr.microsoft.com/devcontainers/java:dev-21-jdk-bullseye` as the container image. The difference is `mcr.microsoft.com/devcontainers/universal:2-focal` is `~10GB` and the `dev-21-jdk-bullseye` image is `~1GB`.
- Setting `workspaceFolder` in the `devcontainer.json` opens to just the directory specified, not the root of the project.
- Adding the Gradle wrapper means that `gradle` doesn't need to be installed. Which is not installed in the image used for this DevContainer.

## Important

I also specified:
```json
	"hostRequirements": {
		"cpus": 4
	},
```

This was due to 2 CPUs being incompatible with the Java devcontainer in use. I noticed this problem with the current Java configuration and decided to change it for this implementation.

holtkampjs committed:
Added java configuration

holtkampjs committed:
Fixed workspace selection

holtkampjs committed:
Fixed workspaceFolder and container name

holtkampjs committed:
Renamed devcontainer

holtkampjs committed:
Created new directory for Java 21 devcontainer

holtkampjs committed:
Removed trailing whitespace

holtkampjs committed:
Removed java21-gradle/Makefile

holtkampjs committed:
Added setting for VSCode Java Null detection

holtkampjs committed:
Updated readme to use ./gradlew

holtkampjs committed:
Defined host requirements

holtkampjs committed:
Removed unused markdown

holtkampjs committed:
Added to description of container

holtkampjs committed:
Added Makefile for simplicity with working with verify pipeline

kylehoehns commented:
This is awesome! Thanks @holtkampjs!

I want to merge this change but there are a couple things I think we want to talk through first.

1. What do you think about making this the default Java project instead of making a new "Java 21 - Gradle" project? Reason being is that I'd like to try and keep things as simple as possible for a few reasons. First, if I'm coming in to do an interview and I know I'm doing it in Java I could see myself being confused on what Java I'd want to select. Second, I've been trying to keep dependencies up to date and I'm not sure it's going to be as easy to keep dependencies updated if we've got two Java projects with roughly the same dependencies. I personally don't see much of a reason that this couldn't just be the default if we think this is a better overall experience.
2. I think you ran into this but we have a [verification step](https://github.com/sourceallies/interviews/blob/main/.github/workflows/verify.yml) that runs on PRs to make sure we can install dependencies and run tests on every project in the repo. That verification workflow is allowing me to be confident that when I'm merging dependency updates we didn't accidentally break something. With this new project, we're now on a different base image than the "universal" one the rest of the app uses. I'm not opposed to making that change, however if we go that route I'm worried that the universal image will drift from the java image we're using in this project and we'll accidentally merge something that doesn't actually work once we launch the app in a Codespace due to the different images we're using. What would you think about altering that GitHub action step to skip this Java project and run a verification on the same image that we run with? Open to other ideas there but I just want to make sure we can continue to be confident in the verify on PRs.

holtkampjs commented:
> This is awesome! Thanks @holtkampjs!
> 
> I want to merge this change but there are a couple things I think we want to talk through first.
> 
> 1. What do you think about making this the default Java project instead of making a new "Java 21 - Gradle" project? Reason being is that I'd like to try and keep things as simple as possible for a few reasons. First, if I'm coming in to do an interview and I know I'm doing it in Java I could see myself being confused on what Java I'd want to select. Second, I've been trying to keep dependencies up to date and I'm not sure it's going to be as easy to keep dependencies updated if we've got two Java projects with roughly the same dependencies. I personally don't see much of a reason that this couldn't just be the default if we think this is a better overall experience.
> 2. I think you ran into this but we have a [verification step](https://github.com/sourceallies/interviews/blob/main/.github/workflows/verify.yml) that runs on PRs to make sure we can install dependencies and run tests on every project in the repo. That verification workflow is allowing me to be confident that when I'm merging dependency updates we didn't accidentally break something. With this new project, we're now on a different base image than the "universal" one the rest of the app uses. I'm not opposed to making that change, however if we go that route I'm worried that the universal image will drift from the java image we're using in this project and we'll accidentally merge something that doesn't actually work once we launch the app in a Codespace due to the different images we're using. What would you think about altering that GitHub action step to skip this Java project and run a verification on the same image that we run with? Open to other ideas there but I just want to make sure we can continue to be confident in the verify on PRs.

Addressing 1)

I'm happy with merging the changes to `java21-gradle/` with what exists in the `java/` directory. The reason I created a new directory was in case someone was attached to the current configuration of the `java/` directory for their "interview flow". I figured some specific naming could be useful for "Java 18 Maven", "Python 3.9 hatch", etc. where custom configurations could be an option since there are so many options available for language versions and build tools that may be incompatible in the same directory.

Addressing 2)

I did bump up against how `Makefile` and the mirrored instructions across the subprojects in this repo depend on the `Makefile` for the Verify step. Initially, I had deleted the `Makefile` since I didn't figure it mattered to the users of the directory I created. 

I noticed using a different image than the `universal` one could cause issues. I was happy to see in the latest passing checks that using the Gradle Wrapper file still succeeds. The Java image the DevContainer I added doesn't include `gradle`.

Seeing as a different image is being used in CI, I did consider that as an issue. What I was debating was a subsequent PR utilizing a matrix in the GitHub Action to use the `devcontainer` command for running and testing the commands for a given subdirectory. `devcontainer` is a NPM package that I've been using for testing out DevContainers locally. I figured it could be a route to open up variation across the containers defined for our `interviews` repo.

holtkampjs commented:
I was also looking at "how to take your interview home" aka how can candidates only take what they did with them. 

The work in progress script I have looks like:
```bash
#!/usr/bin/env bash

# Branch with changes from pairing session
git switch <branch-name>

# Ignore other directories
git filter-branch --subdirectory-filter <directory-used> -- --all

# Add remote to push changes to
git remote add <remote-name> <candidates-new-upstream-repository-url>

# Push changes to new remote repository
git push -u <remote-name> <branch-name>:<upstream-branch>
```
